File: crates/ga-parser/tests/fixtures/ruby-tiny/sinatra-mini/user_service.rb
require_relative 'user_repository'

module App
  class UserService
    def initialize(repo)
      @repo = repo
    end

    def login(id, password)
      user = @repo.find_by_id(id)
      return nil unless user
      user.authenticate(password) ? user : nil
    end

    def register(name, email)
      user = User.new(name, email)
      @repo.save(user)
      user
    end
  end
end

File: crates/ga-parser/tests/fixtures/php-tiny/symfony-mini/UserService.php
<?php

namespace App\Service;

use App\Entity\User;
use App\Repository\UserRepository;
use App\Util\{Logger, Cache};
use App\Auth\AuthService as AuthSvc;
use function strlen;
use const PHP_INT_MAX;

class UserService
{
    #[Required]
    public UserRepository $repo;

    #[ORM\Column(type: 'string')]
    public string $name;

    #[Required]
    #[Lazy]
    public Logger $logger;

    public function getUser(int $id): ?User
    {
        return $this->repo->findById($id);
    }

    public function register(string $name): User
    {
        $user = new User(PHP_INT_MAX, $name);
        Cache::warm($user);
        AuthSvc::grantDefault($user);
        return $this->repo->save($user);
    }

    public function nullsafeLookup(?UserRepository $maybe, int $id): ?User
    {
        return $maybe?->findById($id);
    }

    public function bareCall(string $s): int
    {
        return strlen($s);
    }
}
